IdentityCore/src/oauth2/ciam/MSIDCIAMOauth2Factory.m

```diff
@@ -0,0 +1,152 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+#if !EXCLUDE_FROM_MSALCPP
+
+#import "MSIDCIAMOauth2Factory.h"
+#import "MSIDCIAMTokenResponse.h"
+#import "MSIDOauth2Factory+Internal.h"
+#import "MSIDAccessToken.h"
+#import "MSIDIdToken.h"
+#import "MSIDCIAMAuthority.h"
+
+@implementation MSIDCIAMOauth2Factory
+
++ (MSIDProviderType)providerType
+{
+    return MSIDProviderTypeCIAM;
+}
+
+#pragma mark - Helpers
+
+- (BOOL)checkResponseClass:(MSIDCIAMTokenResponse *)response
+                   context:(id<MSIDRequestContext>)context
+                     error:(NSError **)error
+{
+    if (![response isKindOfClass:[MSIDCIAMTokenResponse class]])
+    {
+        if (error)
+        {
+            NSString *errorMessage = [NSString stringWithFormat:@"Wrong token response type passed, which means wrong factory is being used (expected MSIDCIAMTokenResponse, passed %@", response.class];
+
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, errorMessage, nil, nil, nil, context.correlationId, nil, YES);
+        }
+
+        return NO;
+    }
+
+    return YES;
+}
+
+#pragma mark - Response
+
+- (MSIDCIAMTokenResponse *)tokenResponseFromJSON:(NSDictionary *)json
+                                        context:(__unused id<MSIDRequestContext>)context
+                                          error:(NSError **)error
+{
+    return [[MSIDCIAMTokenResponse alloc] initWithJSONDictionary:json error:error];
+}
+
+- (MSIDCIAMTokenResponse *)tokenResponseFromJSON:(NSDictionary *)json
+                                   refreshToken:(MSIDBaseToken<MSIDRefreshableToken> *)token
+                                        context:(__unused id<MSIDRequestContext>)context
+                                          error:(NSError * __autoreleasing *)error
+{
+    return [[MSIDCIAMTokenResponse alloc] initWithJSONDictionary:json refreshToken:token error:error];
+}
+
+- (BOOL)verifyResponse:(MSIDCIAMTokenResponse *)response
+               context:(id<MSIDRequestContext>)context
+                 error:(NSError * __autoreleasing *)error
+{
+    if (![self checkResponseClass:response context:context error:error])
+    {
+        return NO;
+    }
+
+    return [super verifyResponse:response context:context error:error];
+}
+
+#pragma mark - Tokens
+
+- (BOOL)fillAccount:(MSIDAccount *)account
+       fromResponse:(MSIDCIAMTokenResponse *)response
+      configuration:(MSIDConfiguration *)configuration
+{
+    if (![self checkResponseClass:response context:nil error:nil])
+    {
+        return NO;
+    }
+
+    BOOL result = [super fillAccount:account fromResponse:response configuration:configuration];
+
+    if (!result)
+    {
+        return NO;
+    }
+
+    // CIAM should return preferred_username claim
+    if ([NSString msidIsStringNilOrBlank:response.idTokenObj.preferredUsername])
```

Comment:
Is this extra code on top of AAD?

Reply:
yes, it was unnecessary. removed in most recent commit